src/sx/src/StyleCollector.js

```diff
@@ -1,94 +1,106 @@
-        (typeof styleValue === 'number' || typeof styleValue === 'string')
-      ) {
-        const hash = hashStylePair(property, styleValue, pseudo, this.atRuleName);
+  collect(baseStyleSheet: {|
+    +[sheetName: string]: $FlowFixMe,
+  |}): {| +hashRegistry: HashRegistryType, +styleBuffer: StyleBufferType |} {
+    const hashRegistry: HashRegistryType = new Map();
 
-        addToSheetDefinitions(className ?? '', property, hash, pseudo);
+    const traverse = (styleSheetName, styleSheetObject, styleBuffer, hashSeed = '') => {
+      // we need to iterate it - there might be pseudo classes, at rules or leaf types
+      for (const maybeName of Object.keys(styleSheetObject)) {
+        const maybeValue = styleSheetObject[maybeName];
+        if (typeof maybeValue === 'number' || typeof maybeValue === 'string') {
+          // basic leaf type
+          const node = new StyleCollectorNode(maybeName, maybeValue, hashSeed);
+          styleBuffer.add(node);
 
-        this.styles.set(
-          hash,
-          new StyleNode(transformStyleName(property), transformValue(property, styleValue), pseudo),
-        );
-      } else if (typeof styleValue === 'object' && styleValue != null) {
-        if (property.startsWith('@media') || property.startsWith('@supports')) {
-          // Media queries are represented by their own style collector, this way we can have as
-          // many nested levels as needed
-          const collector =
-            ((this.styles.get(property): any): StyleCollector) ?? new StyleCollector(property);
-          this.styles.set(property, collector);
-          collector.addStyles(styleValue, hashedSheetDefinitions, className);
-        } else if (property.startsWith('@')) {
-          throw new Error(sprintf('Unsupported at rule "%s"', property));
+          // add record to the has registry
+          const hashRegistryKey = `${maybeName}${hashSeed}`;
+          if (hashRegistry.has(styleSheetName)) {
+            hashRegistry.get(styleSheetName)?.set(hashRegistryKey, node.getHash());
+          } else {
+            hashRegistry.set(styleSheetName, new Map([[hashRegistryKey, node.getHash()]]));
+          }
+        } else if (maybeName.startsWith(':')) {
+          // pseudo type
+          const nodes = traverse(styleSheetName, maybeValue, new Set([]), maybeName);
+          styleBuffer.add(new StyleCollectorPseudoNode(maybeName, nodes));
+        } else if (maybeName.startsWith('@media') || maybeName.startsWith('@supports')) {
+          // at rule type
+          const nodes = traverse(styleSheetName, maybeValue, new Set([]), maybeName);
+          styleBuffer.add(new StyleCollectorAtNode(maybeName, nodes));
         } else {
-          const isPseudo = property.startsWith(':');
-          iterateEntries(
-            Object.entries(styleValue),
-            isPseudo ? property : pseudo,
-            isPseudo ? className : property,
-          );
+          throw new Error(`Unsupported rule "${maybeName}"`);
```

Comment:
I am wondering if throwing is a little bit too harsh, this could cause runtime errors, would maybe throwing only when __DEV__ === true be enough?

Reply:
improved